# Data Analysis Assignment 2: Customer Service Data Solutions

## Business Issue and Understanding

## Getting Started
You will have two csv files to work 

In [370]:
import pandas as pd
import numpy as np

north = pd.read_csv("NorthCallCenter.csv")
south = pd.read_csv("SouthCallCenter.csv")

## Exploratory Data Analysis



In [371]:
north.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 245 entries, 0 to 244
Data columns (total 15 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 --------------  -----  
 0   Unnamed: 0             245 non-null    int64  
 1   Branch                 245 non-null    object 
 2   Call Purpose           245 non-null    object 
 3   Time Block             245 non-null    object 
 4   Incoming or Outgoing   245 non-null    object 
 5   Queue                  245 non-null    object 
 6   Rep ID                 245 non-null    object 
 7   Sale                   245 non-null    object 
 8   Lost Call              245 non-null    int64  
 9   Abandoned              245 non-null    int64  
 10  Busy Minutes           245 non-null    int64  
 11  Not Ready Minutes      245 non-null    int64  
 12  Incoming Wait Time     82 non-null     float64
 13  During Call Wait Time  245 non-null    int64  
 14  Calls                  245 non-null    int64  
dtypes: flo

In [372]:
north.head()

Unnamed: 0 Branch     Call Purpose Time Block Incoming or Outgoing Queue  \
0           0  North        Complaint    9:00 AM             Incoming     B   
1           1  North    Sales Support   11:00 AM             Incoming     A   
2           2  North    Sales Support   11:00 AM             Incoming     B   
3           3  North    Sales Support    4:00 PM             Incoming     B   
4           4  North  Product Support   12:00 PM             Incoming     B   

  Rep ID Sale  Lost Call  Abandoned  Busy Minutes  Not Ready Minutes  \
0  Brent   NO          0          1             9                  2   
1    Cam   NO          0          0            11                  1   
2   Todd   NO          0          0             9                  3   
3  Brent   NO          0          0            11                  2   
4  Brent   NO          0          0             8                  2   

   Incoming Wait Time  During Call Wait Time  Calls  
0                 1.0                      4      1  
1                 1.0                      3      1  
2                 1.0                      3      1  
3                 1.0                      4      1  
4                 1.0                      4      1

1. How many reps in each branch?
1. What is the average busy minutes, not ready minutes, incoming call wait time, during call wait time for each branch?
1. What is the number of calls taken in for each time block?

<span style="background-color:orangered; color:orangered;">- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -</span>

In [373]:
# Question 1: 
north_number_reps = north["Rep ID"].nunique()
south_number_reps = south["Rep ID"].nunique()


print("There are {0} reps in the north branch, and {1} reps in the south branch.".format(north_number_reps, south_number_reps))

There are 9 reps in the north branch, and 11 reps in the south branch.


In [374]:
north_reps = north["Rep ID"].value_counts() #<-- this could work
north_reps

Brent     37
Lilly     30
Todd      29
Duke      29
Joe       26
Xander    24
Cam       24
Amanda    23
Andy      23
Name: Rep ID, dtype: int64

In [375]:
# Question 2:
north.describe()

Unnamed: 0   Lost Call   Abandoned  Busy Minutes  Not Ready Minutes  \
count  245.000000  245.000000  245.000000    245.000000         245.000000   
mean   122.000000    0.020408    0.028571      9.987755           1.914286   
std     70.869599    0.141681    0.166940      1.403980           0.771416   
min      0.000000    0.000000    0.000000      8.000000           1.000000   
25%     61.000000    0.000000    0.000000      9.000000           1.000000   
50%    122.000000    0.000000    0.000000     10.000000           2.000000   
75%    183.000000    0.000000    0.000000     11.000000           3.000000   
max    244.000000    1.000000    1.000000     12.000000           3.000000   

       Incoming Wait Time  During Call Wait Time  Calls  
count           82.000000             245.000000  245.0  
mean             3.048780               2.967347    1.0  
std              1.456382               0.843827    0.0  
min              1.000000               2.000000    1.0  
25%              2.000000               2.000000    1.0  
50%              3.000000               3.000000    1.0  
75%              4.000000               4.000000    1.0  
max              5.000000               4.000000    1.0

In [376]:
south.describe()

Unnamed: 0   Lost Call   Abandoned  Busy Minutes  Not Ready Minutes  \
count  314.000000  314.000000  314.000000     314.00000         314.000000   
mean   156.500000    0.035032    0.012739      10.05414           1.910828   
std     90.788215    0.184154    0.112324       1.40751           0.860016   
min      0.000000    0.000000    0.000000       8.00000           1.000000   
25%     78.250000    0.000000    0.000000       9.00000           1.000000   
50%    156.500000    0.000000    0.000000      10.00000           2.000000   
75%    234.750000    0.000000    0.000000      11.00000           3.000000   
max    313.000000    1.000000    1.000000      12.00000           3.000000   

       Incoming Wait Time  During Call Wait Time  Calls  
count          126.000000             314.000000  314.0  
mean             3.000000               3.076433    1.0  
std              1.308434               0.819424    0.0  
min              1.000000               2.000000    1.0  
25%              2.000000               2.000000    1.0  
50%              3.000000               3.000000    1.0  
75%              4.000000               4.000000    1.0  
max              5.000000               4.000000    1.0

In [377]:
# Question 3: 
north_calls_per_time_block = north["Time Block"].value_counts()
north_calls_per_time_block

12:00 PM    45
10:00 AM    43
2:00 PM     35
3:00 PM     30
4:00 PM     28
11:00 AM    25
1:00 PM     17
5:00 PM     17
9:00 AM      5
Name: Time Block, dtype: int64

In [378]:
south_calls_per_time_block = south["Time Block"].value_counts()
south_calls_per_time_block

12:00 PM    75
10:00 AM    56
3:00 PM     43
11:00 AM    31
2:00 PM     30
5:00 PM     26
4:00 PM     25
1:00 PM     23
9:00 AM      5
Name: Time Block, dtype: int64

## Cleaning Your Data

1. Incoming wait time is null in many places.  Do all of those places correspond to outgoing calls?
    - keep this
1. In the north data set, there are 2 separate "YES" values for a sale.  Why and how do you rectify the issue?
    - took me forever, but i figured it out!  (stupid extra space)

1. Any other ideas??  If we keep this small, we could add 1 or 2 more questions to the manipulation section?

<span style="background-color:orangered; color:orangered;">- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -</span>

In [379]:
# Question 1:
north_dup_check = north.duplicated().sum()
# north_dup_check

In [380]:
south_dup_check = south.duplicated().sum()
# south_dup_check

In [381]:
north_rep_check = north.groupby(["Rep ID"]).agg(pd.Series.mode)
# north_rep_check 

In [382]:
combo = north.append(south)
# combo.head()

In [383]:
combo_rep_check = combo.groupby(["Rep ID"]).agg(pd.Series.mode)
# combo_rep_check 

In [384]:
check = combo.duplicated
# check

In [385]:
N_rep_check = north.groupby(["Rep ID"]).agg(pd.Series.mode)
# N_rep_check 

In [386]:
check1 = combo.drop_duplicates()
# check1


In [387]:
Ncheck = north["Rep ID"].unique
# Ncheck

In [388]:
Scheck = south["Rep ID"].unique()
# Scheck

In [389]:
S_rep_check = south.groupby(["Rep ID"]).agg(pd.Series.mode)
# S_rep_check 

In [390]:
combo_queue_check = combo[["Queue", "Rep ID"]]
# combo_queue_check

<span style="background-color:orangered; color:orangered;">- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -</span>

In [391]:
# Question 2:
north_in_out_check = north["Incoming or Outgoing"].value_counts()
north_in_out_check

Outgoing    163
Incoming     82
Name: Incoming or Outgoing, dtype: int64

In [392]:
north_iwt = north["Incoming Wait Time"].value_counts(dropna=False)
north_iwt

NaN    163
4.0     20
5.0     17
1.0     17
2.0     16
3.0     12
Name: Incoming Wait Time, dtype: int64

ABOVE: Just looking at numbers here -- should be 163 outgoing, and there are 163 NaN, but not exact confirmation

BELOW: created table based on incoming/outgoing and then counted incoming wait time.  more exact.

Resource: https://pandas.pydata.org/docs/getting_started/intro_tutorials/03_subset_data.html

In [393]:
south_in_out_check = south["Incoming or Outgoing"].value_counts()
south_in_out_check

Outgoing    188
Incoming    126
Name: Incoming or Outgoing, dtype: int64

In [394]:
south_out_table = south[south["Incoming or Outgoing"] == "Outgoing"]
south_out_table["Incoming Wait Time"].value_counts()


Series([], Name: Incoming Wait Time, dtype: int64)

In [395]:
south_in_table = south[south["Incoming or Outgoing"] == "Incoming"]
south_in_table["Incoming Wait Time"].value_counts()

4.0    30
3.0    29
2.0    28
1.0    20
5.0    19
Name: Incoming Wait Time, dtype: int64

In [396]:
# Question 3:

In [397]:
yes_check = north["Sale"].value_counts()
yes_check

NO      234
YES      10
YES       1
Name: Sale, dtype: int64

In [398]:
new_yes = north["Sale"].str.replace("YES ", "YES")
new_yes.value_counts() # check - 11 yes
north["Sale"] = new_yes

In [399]:
# no_table = north[north["Sale"] != "YES"]
# no_table["Sale"].value_counts()
# no_table1 = no_table[no_table["Sale"] != "NO"]
# no_table1["Sale"].value_counts()
# no_table1.head()
# no_table1["Sale"].replace("YES ", "YES")

#how i found the row, and then looked at the damn csv to see what the deal was


In [400]:
north["Sale"].value_counts()

NO     234
YES     11
Name: Sale, dtype: int64

<span style="background-color:orangered; color:orangered;">- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -</span>

In [401]:
# Question 4:

In [405]:
queue_check = north[["Queue", "Rep ID"]].isnull().sum()
queue_check

Queue     0
Rep ID    0
dtype: int64

## Data Manipuation

Merge the 2 data sets together to comapre south v north AND company wide trends

1. Group by Rep ID and sum the structure and sort by calls to determine which rep in each branch has the highest number of calls. 
df.groupby(rep).sum()
1. What is the average call abaondonment rate for each branch?
1. What is the percentage of call answered within 2 seconds?  
1. What is the average speed of answering?

## Visualization

For each chart you created, eplain why you chose the chart style.

1. Abandonment Rate per Queue
1. What is the service level and average speed of answer per each rep?
1. For each type of call purpose, how many calls are outgoing vs incoming?  
 

## Submit Your Work

## Bonus Mission

1. What are the time blocks of these call purposes? - groupby and plot